src/maxtree.jl

```diff
@@ -0,0 +1,881 @@
+   > https://doi.org/10.1109/ICIP.2007.4379949
+3. Najman, L., & Couprie, M. (2006). *Building the component tree in quasi-linear time*. IEEE Transactions on Image Processing, 15(11), 3531-3539.
+   > https://doi.org/10.1109/TIP.2006.877518
+4. Carlinet, E., & Geraud, T. (2014). *A Comparative Review of Component Tree Computation Algorithms*. IEEE Transactions on Image Processing, 23(9), 3885-3895.
+   > https://doi.org/10.1109/TIP.2014.2336551
+"""
+struct MaxTree{N,A}
+    """
+    Axes of the source image.
+    """
+    axes::A
+
+    """
+    If `false`, the tree is constructed from the smallest to the largest values,
+    the tree root being the smallest (darkest) pixel.
+    Otherwise, if `true`, the direction is from large to small values.
+    """
+    rev::Bool
+
+    """
+    Array of the same shape as the source image.
+    Each value is the linear index of the parent pixel in the max-tree.
+    """
+    parents::Array{Int,N}
+
+    """
+    The order of the elements in the tree, from root to leaves.
+    Each element corresponds to the linear index of the pixel in the image.
+    A parent of a pixel always comes before the pixel itself, i.e.
+    if index `i` comes before `j`, then `j`-th pixel cannot be
+    the parent of `i`-th pixel.
+    """
+    traverse::Vector{Int}
+
+    # create uninitialized MaxTree for image
+    MaxTree{N}(image::AbstractArray{<:Any, N}, rev::Bool) where N =
+        new{N, typeof(axes(image))}(axes(image), rev, similar(image, Int),
+                                    Vector{Int}(undef, length(image)))
+end
+
+Base.ndims(::Type{<:MaxTree{N}}) where N = N
+Base.ndims(maxtree::MaxTree) = ndims(typeof(maxtree))
+Base.axes(maxtree::MaxTree) = maxtree.axes
+Base.length(maxtree::MaxTree) = length(maxtree.parents)
+Base.size(maxtree::MaxTree) = size(maxtree.parents)
+Base.:(==)(a::MaxTree{N, A}, b::MaxTree{N, A}) where {N, A} =
+    (a.rev == b.rev) && (a.axes == b.axes) &&
+    (a.parents == b.parents) && (a.traverse == b.traverse)
+Base.isequal(a::MaxTree{N, A}, b::MaxTree{N, A}) where {N, A} =
+    isequal(a.rev, b.rev) && isequal(a.axes, b.axes) &&
+    isequal(a.parents, b.parents) && isequal(a.traverse, b.traverse)
+
+"""
+    root_index(maxtree::MaxTree) -> Int
+
+Linear index of the root pixel.
+"""
+root_index(maxtree::MaxTree) = maxtree.traverse[1]
+
+# Gets the path to the root of the max-tree (defined by `ancestors`) starting from
+# `node`. The indices of the nodes along the path are stored into `path`.
+# Returns the index of the root
+@inline function rootpath!(path::Vector{Int},
+                           ancestors::AbstractArray{<:Integer},
+                           node::Integer)
+    empty!(path)
+    p = node
+    @label next
+    push!(path, p)
+    @inbounds q = ancestors[p]
+    if p != q
+        p = q
+        @goto next
+    end
```

Comment:
This is a most performance-critical section of the algorithm.
Doing it like this instead of a `while` loop helps to avoid repeating `push!()` before/after loop, so it may help the compiler to generate efficient code.
IMHO the code structure here is quite clear. I'd rather keep it as is.